### DISCUSSION TITLE
json tags missing

### DISCUSSION THREAD

**Discussion by williamgorden** [go-webauthn/webauthn#245] [Q&A]
Good morning.
Congratulations on this wonderful module.
I am working on a module for a web app where I want to login with my hardware key.
It is enough for me to store user data in a json file.
Here in the repo structs like Credentials have json tags.
I ran "go get" to easily include go-webauthn in my project.
I also ran go mod vendor to better be able to inspect the code in VS-Code.
However, the copy in the vendor directory has no json tags!!
json.Unmarshal no longer works.
Why are there no json tags in the structs?
What am I missing?

**Comment by james-d-elliott** [comment]
Which structs are you having the issue with?

**Reply by williamgorden** [reply]
Hi, that was quick!
Thank you.
My user-struct has a field:
Credentials []webauthn.Credential
It doesn't get unmarshaled.
The Credential struct here (in the repo) in the code file credential.go has the following field:
ID []byte `json:"id"` (example field)
All the Fields have json-tags like this.
However after running: go get github.com/go-webauthn/webauthn
and also: go mod vendor
In the same file in my vendor directory in the Credential struct all the fields are missing json-tags:
ID []byte (example field)
This seems strange.
When I try to unmarshal this from a json-file, it doesn't work.
Here are the bytes from the file:
datastore bytes:
{
    "Users": {
        "bill@gorden.dev": {
            "id": "3302708262600056270",
            "name": "bill@gorden.dev",
            "display_name": "bill",
            "credentials": [
                {
                    "ID": "ZVF6uhePxc2HDFFS3xc0H6tW66GLXcM476Oohw2ebdvxyvm8LuPZUCVUMb0N+kj93UM1+D9Esp6MwvsIkOfwmbnockdIBc77snAZ0BlsaKh3L5UUI+Wj9LVeE+KxB2ac",
                    "PublicKey": "pQECAyYgASFYIKCGiZp8g1NVW36wTmcwcKc1lH+0DKZOuHjrdgojSfKWIlggKc9UR0jR4rKJ8qoNSFCZUH4zN0LJFqinSf+U/X8kiVg=",
                    "AttestationType": "none",
                    "Transport": null,
                    "Authenticator": {
                        "AAGUID": "AAAAAAAAAAAAAAAAAAAAAA==",
                        "SignCount": 1123,
                        "CloneWarning": false
                    }
                }
            ]
        }
    }
}

Here is the unmarshaled struct:
{map[bill@gorden.dev:{3302708262600056270 bill@gorden.dev bill  [{[101 81 122 186 23 143 197 205 135 12 81 82 223 23 52 31 171 86 235 161 139 93 195 56 239 163 168 135 13 158 109 219 241 202 249 188 46 227 217 80 37 84 49 189 13 250 72 253 221 67 53 248 63 68 178 158 140 194 251 8 144 231 240 153 185 232 114 71 72 5 206 251 178 112 25 208 25 108 104 168 119 47 149 20 35 229 163 244 181 94 19 226 177 7 102 156] [165 1 2 3 38 32 1 33 88 32 160 134 137 154 124 131 83 85 91 126 176 78 103 48 112 167 53 148 127 180 12 166 78 184 120 235 118 10 35 73 242 150 34 88 32 41 207 84 71 72 209 226 178 137 242 170 13 72 80 153 80 126 51 55 66 201 22 168 167 73 255 148 253 127 36 137 88] none [] {[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0] 1123 false}}]}]}
loaded datastore:
{map[]}

I edited this into the following to make it readable for you:
{
    map[
        bill@gorden.dev:
        {
            2508547699174237232 
            bill@gorden.dev 
            bill
            [
                credential: {
                    id: [224 197 213 48 172 60 253 128 131 178 66 88 199 247 150 69 93 30 156 50 25 1 32 131 143 72 194 43 50 89 183 74 228 145 200 108 80 8 110 36 195 41 143 11 150 122 77 70 171 239 153 234 31 220 111 52 132 74 22 1 57 180 228 222 177 17 117 80 41 95 94 161 214 31 197 82 13 54 0 245 217 240 24 26 12 136 203 80 64 205 61 100 186 197 90 28] 
                    publicKey: [165 1 2 3 38 32 1 33 88 32 183 20 120 39 125 237 18 237 13 187 157 246 153 18 229 241 42 247 108 221 138 5 162 214 103 79 239 24 232 51 186 231 34 88 32 23 157 174 167 198 58 111 201 105 39 52 81 154 125 141 56 111 96 196 37 183 55 197 92 192 240 114 206 42 136 89 231] 
                    attestationType: none 
                    transport: [] 
                    {
                        [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0] 
                        961 
                        false
                    }
                }
            ]
        }
    ]
}

Everything above the credential is my User struct, where the fields have json-tags.
I added labels here in the credential part above, so it would be easier to see what was missing.
Registering the user and the hardware key works great!
I save this in a json-file to be able to have a user-database (simplified).
The data is actually there, it just doesn't make it into the struct, which is in turn useless. I get an error, that the user cannot be found.

Sorry if this is too verbose.

Thanks for the help.

I think everything will work great if the json-tags are there.

**Reply by williamgorden** [reply]
Here is a section of the file _credential.go_ retrieved with go mod vendor:

`// Credential contains all needed information about a WebAuthn credential for storage
type Credential struct {
	// A probabilistically-unique byte sequence identifying a public key credential source and its authentication assertions.
	ID []byte
	// The public key portion of a Relying Party-specific credential key pair, generated by an authenticator and returned to
	// a Relying Party at registration time (see also public key credential). The private key portion of the credential key
	// pair is known as the credential private key. Note that in the case of self attestation, the credential key pair is also
	// used as the attestation key pair, see self attestation for details.
	PublicKey []byte
	// The attestation format used (if any) by the authenticator when creating the credential.
	AttestationType string
	// The transport types the authenticator supports.
	Transport []protocol.AuthenticatorTransport
	// The Authenticator information for a given certificate
	Authenticator Authenticator
}
`
Here there are no json tags...

**Comment by james-d-elliott** [comment]
Seems very strange. The `encoding/json` package will marshall all exported fields excluding those with the tags `json:"-"`. Show your user struct?

**Reply by williamgorden** [reply]
`type User struct {
	ID          string                `json:"id"`
	Name        string                `json:"name"`
	DisplayName string                `json:"display_name"`
	Icon        string                `json:"icon,omitempty"`
	Credentials []webauthn.Credential `json:"credentials,omitempty"`
}`

github uploads don't support go-files...

**Reply by james-d-elliott** [reply]
What version of go and the package are you downloading?

**Reply by williamgorden** [reply]
I just updated go to version 1.22.4
I just redid the package version in my go.mod to:
`require github.com/go-webauthn/webauthn v0.10.2`
I repeated "go mod vendor"
The json tags are now present.
The package version was wrong.

After running again...
Now registering is not working...
I need to work on this more before we go any further with this.
Sorry...

What do we do with this QA?

**Reply by williamgorden** [reply]
Hello Mr. Elliot.
Registering and logging in works now.

I found another Problem in the Javascript part.
My decodeBuffer and encodeBuffer functions were incorrect.
I have been using some of the code from a Duo-labs example.

I like that example because its very simple, doesn't use Frameworks and docker, etc.
I can learn more about how this works without reading tons of code.

I bet the webauthn-example works well, but its a lot to go through.
My own projects implement golang-http-templates that I serve up from the new mux in the standard library.

Thankyou for your patience yesterday.

The motto of this story is: check everything more exactly before starting QAs like this.

**Reply by james-d-elliott** [reply]
Nice! Glad you got it working. I have not updated the example in quite some time, so it may not work. 

For reference I personally use [@simplewebauthn/browser](https://www.npmjs.com/package/@simplewebauthn/browser) which handles most of the JS.

**Reply by williamgorden** [reply]
Thanks for the encouragement.
The learning curve is steep for me this time.
It makes more sense every hour I spend on this.
My website (with a short biography) is in my profile, if you're interested.
Have a good day.
